add dynamic enum Tutorial

LukasBoll opened the pull request:
closes #270

LukasBoll committed:
add dynamic enum Tutorial

Co-authored-by: Stefan Dirix <sdirix@eclipsesource.com>
closes #270

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,260 @@
+Imagine a form where users need to provide their location by selecting a country, a region and a city.
+The options for countries and regions are fetched from an API.
+The available regions depend on the selected country.
+To address those requirements, we'll create custom renderers for the country and region.
```
```suggestion
To address those requirements, we'll create custom renderers for country and region.
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,260 @@
+The options for countries and regions are fetched from an API.
+The available regions depend on the selected country.
+To address those requirements, we'll create custom renderers for the country and region.
+The example is also implemented in the [react-seed](https://github.com/eclipsesource/jsonforms-react-seed) app.
```
```suggestion
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,260 @@
+To begin, let's introduce the corresponding JSON schema.
+We have created an object with properties for country, region, and city.
+In our example, the schema also includes a property `x-url`, which specifies the entry point of the corresponding API.
+Both `country` and `region` have a property `endpoint`, indicating the endpoint from which the data should be fetched.
```
```suggestion
Both `country` and `region` have a property `x-endpoint`, indicating the endpoint from which the data should be fetched.
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
````diff
@@ -0,0 +1,260 @@
+```js
+export const APIContext = React.createContext(new API(url));
+
+const App = () =>{
+
+  ...
+  <JsonForms/>
+}
````
We need to use the context in the JSX somewhere

LukasBoll replied:
Thanks, I added your suggestions. The context is used in the country and region renderer, but I also added an example on how to access the context right below this chapter.

sdirix replied:
Ah I see, you're relying on the default context and skip the
```tsx
const APIService = new API(url);
// ...
<APIContext.Provider value={APIService}>
  <JsonForms />
</APIContext.Provider />
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,260 @@
+
+### The Country Renderer
+
+The core of the country renderer is a dropdown, we can reuse the MaterialEnumControl from the material-renderer set.
```
```suggestion
The core of the country renderer is a dropdown, therefore we can reuse the MaterialEnumControl from the React Material renderer set.
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,260 @@
+#### Accessing Schema Data
+
+The `endpoint` and `dependent` fields can be obtained from the schema object provided to the custom renderer via JSON Forms.
+Since these fields are not part of the standard JSON schema type in JSON Forms, we must add them to the schema´s interface and access them as follows:
```
```suggestion
Since these fields are not part of the standard JSON schema type in JSON Forms, we must add them to the schema's interface and access them as follows:
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,260 @@
+#### The Country Renderer
+
+The country renderer uses the `APIContext` to query the API and fetch the available options.
+We utilize the `useEffect` hook to reload new options, if API changes.
```
```suggestion
We utilize the `useEffect` hook to initialize the options.
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,260 @@
+    api.get(endpoint).then((result) => {
+      setOptions(result);
+    });
+  }, [api, endpoint]);
```
```suggestion
  useEffect(() => {
    api.get(endpoint).then((result) => {
      setOptions(result);
    });
  }, []);
```

sdirix reviewed (commented):
I like it!

Please add a small section at the end that this behavior can be generalized and extracted into React HOCs for cleaner code.

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
````diff
@@ -0,0 +1,259 @@
+};
+```
+Again we need to create a [create a tester](./custom-renderers#2-create-a-tester) and [register](./custom-renderers#3-register-the-renderer) the new custom renderer.
+A running example of the scenario can also be seen at the [react-seed](https://github.com/eclipsesource/jsonforms-react-seed) app.
````
```suggestion
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
````diff
@@ -0,0 +1,259 @@
+    />
+  );
+};
+```
````
Please add the wrapping code for the renderers, i.e. `withJsonFormsControlProps` etc. here and in the other listings.

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,259 @@
+  endpoint: string;
+};
+
+export const Country = (
```
let's not export the raw renderers
```suggestion
const CountryControl = (
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
```diff
@@ -0,0 +1,259 @@
+#### Initializing the React Context
+
+Now that we have access to the API URL, we can use React Context to make this data available across our renderers.
+React Context allows you to share data globally within your application, enabling components deep in the component tree to access data without needing to pass properties through all parent elements.
```
Link to the official react context documentation of react

```suggestion
React Context allows you to share data deep in the component tree to access data without needing to pass additional properties through the component hierarchy.
```

sdirix commented on content/docs/tutorial/dynamic-enum.mdx:
````diff
@@ -0,0 +1,259 @@
+const api = React.useContext(APIContext);
+```
+
+Changing the context's value will trigger a re-render of components that use it, making it a powerful tool for managing dynamic data.
````
```suggestion
Changing the context's value will trigger a re-render of components that use it.
```

LukasBoll commented:
thank you, I applied the changes from your review!

sdirix committed:
Merge remote-tracking branch 'upstream/master' into dynamicEnumTutorial

sdirix committed:
clean up code examples

sdirix committed:
replace x-dependencies with x-dependents

sdirix committed:
Merge remote-tracking branch 'upstream/master' into dynamicEnumTutorial

lucas-koehler committed:
Merge remote-tracking branch 'upstream/master' into dynamicEnumTutorial

lucas-koehler committed:
Small typo, wording, whitespace fixes

lucas-koehler committed:
cleanup  example code

- correctly use `x-dependent` and `x-endpoint`
- use useJsonFormsControlProps instead of withJsonFormsEnumProps because options are calculated manually

lucas-koehler reviewed (approved):
Looked pretty good to me already. I did a few minor fixes regarding wording and using `x-` prefixed schema properties for endpoint and dependents.
